### DISCUSSION TITLE
UV map auto-orientation

### DISCUSSION THREAD

**Discussion by csparker247** [educelab/volume-cartographer#49] [Ideas]
We would like to auto-orient the UV map with respect to at least the volume (i.e. no frame of reference except the volume shape), but also landmarks in the volume (e.g. top, bottom, front, back, etc.).

The appropriate place to fix UV map auto-orientation for general objects (i.e. without any frame of reference) is in the `FlatteningAlgorithm::orient_uvs_` method so that it's available to all subclasses. That's the base class for all UV algorithms, and all subclasses would need to call it inside their own `compute()` functions. The current implementation (which tried to align the z-axis to the v-axis) is not used because it was pretty unreliable in practice. At the moment, the current UV Map orientation is determined by [the two pinned edges in the OpenABF LSCM code](https://github.com/educelab/OpenABF/blob/develop/include/OpenABF/AngleBasedLSCM.hpp#L84), which is functionally arbitrary.

Auto-orientation of the UV map with respect to landmarks should be implemented as a separate function/class that takes a UV map and landmarks as input and outputs an updated UV map. We really want to auto-orient the text in a readable direction. Once a single piece of text is found and oriented, all other text should be auto-orientable as a result.

**Comment by jrudolph** [comment]
> The current implementation (which tried to align the z-axis to the v-axis) is not used because it was pretty unreliable in practice

Interesting. The algorithm seems to use all points for the optimization which seems to be quite an overspecification of the problem (since we are looking for a single rotation angle, basically).

What worked reasonably well in my tests is sampling N points, and then for each of those look at M points spread on a circle around and then choose the direction with most negative d_z between center and circle point. Then average over the directions found for those N points.

**Reply by csparker247** [reply]
Sounds interesting. I'm happy to review a PR to this effect.

**Comment by spelufo** [comment]
I don't think I'll work on this but here are some thoughts. I see three different goals related to this:

1. Consistency, just being predictable. E.g. orient relative to original axis, but also consistent handedness of the UVN frames at faces.
2. Not wasting space (and compute of ink detection if it doesn't check for this) in the generated layers by orienting the UV to crop the rectangle tightly.
3. Attempting to match the desired coordinate system intrinsic to the data. E.g. U going roughly with horizontal fibers in papyrus scrolls, or parallel to the edge of a page in a book.

Regarding 3 in the case of scrolls again having the core would be handy. U could be roughly aligned with the tangent to circles around the scroll core axis.

**Comment by stephenrparsons** [comment]
I'll work on an implementation tomorrow that is an improvement to the existing one aligning z-axis to v-axis, and will sample fewer points. The main goal will be auto-orientation for Vesuvius Challenge Scroll 1 segments, so that text is consistently aligned correctly. Will be an option to enable it during vc_render.

**Comment by stephenrparsons** [comment]
PR open: https://github.com/educelab/volume-cartographer/pull/57

**Comment by csparker247** [comment]
Solved in #57.
